### PR TITLE
[Mellanox] enhance the mlnx-sfpd init flow  

### PR DESCRIPTION
**Pull request by keboliu**
**- What I did**
Enhance the mlnx-sfpd flow, 3 sdk API are called during mlnx-sfpd init, in some cases these sdk API may not ready and return false, need retry.

**- How I did it**
Add code to retry all the SDK API calls, if failed, wait for some time and call again until exceed the max retry time. 

**- How to verify it**
reboot, config reload and check the init log of mlnx-sfpd

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by liat-grozovik** [comment]
retest vs please

File: platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd [hunk 1]
```diff
@@ -101,34 +100,65 @@ class MlnxSfpd:
...
-        if rc != SX_STATUS_SUCCESS:
-            raise RuntimeError("sx_api_host_ifc_open exited with error, rc {}".format(rc))
-
-        self.user_channel_p.type = SX_USER_CHANNEL_TYPE_FD
-        self.user_channel_p.channel.fd = self.rx_fd_p
+            if not sx_api_open_pass:
+                rc, self.handle = sx_api_open(None)
+                if rc == SX_STATUS_SUCCESS:
+                    sx_api_open_pass = True
+                    retry = 1
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: >retry = 1 [](start = 20, length = 9)

No retry happen, why assign 1?

File: platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd [hunk 2]
```diff
@@ -101,34 +100,65 @@ class MlnxSfpd:
         # open SDK API handle
         # retry at most SX_OPEN_RETRIES times to wait
         # until SDK is started during system startup
+        
         retry = 1
+        sx_api_open_pass = False
```
Comment (qiluo-msft) [inline platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: >sx_api_open_pass [](start = 8, length = 16)

Should we blame the design of SDK API?
1. if api_open() is not ready, we should not start this downstream service at all
2. if api_open() is ready, all other api should be also ready.

File: platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd (hunk 1 above)
Comment (keboliu) [reply platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: you are correct, fixed.

File: platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd (hunk 2 above)
Comment (keboliu) [reply platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: I can not say this is a bug of SDK or by design, this is what we can do so far.

File: platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd [hunk 3]
```diff
@@ -101,34 +100,65 @@ class MlnxSfpd:
...
+                                                          SX_TRAP_ID_PMPE,
+                                                          self.user_channel_p)
+                if rc == SX_STATUS_SUCCESS:
+                    sx_api_host_ifc_trap_id_register_set_pass = True
+                    break
+                else:
+                    log_warning("sx_api_host_ifc_trap_id_register_set failed with error code {}... retrying {}".format(rc,retry))
+                    retry += 1
+                    time.sleep(2 ** retry)
+                    if retry > self.SX_OPEN_RETRIES:
```
Comment (stcheng) [inline platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: `SX_OPEN_RETRIES` is set to 20. Will it be too long to wait for 2 ^ 20 seconds?

Comment (qiluo-msft) [reply platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: Echo Shuotian. It doesn't make sense to sleep 12.1 days and retry. One suggestion is keep (2^retry) until a maximal value.

---
In reply to: [306872036](https://github.com/Azure/sonic-buildimage/pull/3193#discussion_r306872036) [](ancestors = 306872036)

Comment (keboliu) [reply platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd]: set max retry time to 10, max waiting time for each API will be 1024 sec

**Comment by qiluo-msft** [comment]
Liat has some concern and we need follow up discussion.

**Pull request closed without merging**